[Issue #46210] added torch.fx.len() to provide support for len(); added a test case for torch.fx.len()

huiguoo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49532 [Issue #46210] added torch.fx.len() to provide support for len(); added a test case for torch.fx.len()**

Differential Revision: [D25608804](https://our.internmc.facebook.com/intern/diff/D25608804)

huiguoo committed:
[Issue #46210] added torch.fx.len() to provide support for len(); added a test case for torch.fx.len()

[ghstack-poisoned]

jamesr66a commented on torch/fx/__init__.py:
```diff
@@ -90,4 +90,4 @@ def forward(self, x):
 from .graph import Graph
 from .node import Node, map_arg
-from .proxy import Proxy
+from .proxy import Proxy, len
```
It doesn't seem to be causing typechecking issues in this PR, but do you mind adding an import for `len` to `torch/fx/__init__.pyi` as well? There are some situations where importing `torch.fx.foo` where foo is imported in the `__init__.py` but not in the `__init__.pyi` causes issues

jamesr66a commented on torch/fx/proxy.py:
```diff
@@ -153,6 +156,10 @@ def __bool__(self) -> bool:
 
+    def __len__(self):
+        warnings.warn("'builtins.len' is not supported. Replace `builtins.len` with 'torch.fx.len' and try again.")
+        return
```
Instead of warning, this should throw a `RuntimeError`. As it is, this function will return `None`, and I think it will run into the same hard-to-interpret "can't be interpreted as an int" error. Can you make this change and add a test case for the exception case (using `self.assertRaisesRegex` in the test)?

jamesr66a commented on test/test_fx.py:
```diff
@@ -1161,5 +1161,13 @@ def forward(self, x):
+
+        traced = symbolic_trace(LenTest())
+
+
```
nit: extra blank line

And also a practical thing: adding tests to the very end can end up with merge conflicts if someone else adds a test similarly and tries to land it at the same time. I typically add tests somewhere in the middle of the file. (dumb solution, I know, but it works :) )

huiguoo replied:
Very useful, thanks James :)

jamesr66a reviewed (commented):
Great! Some small stuff then I think this should be good to go

huiguoo committed:
Update on "[Issue #46210] added torch.fx.len() to provide support for len(); added a test case for torch.fx.len()"


Differential Revision: [D25608804](https://our.internmc.facebook.com/intern/diff/D25608804)

[ghstack-poisoned]

huiguoo committed:
Update on "[Issue #46210] added torch.fx.len() to provide support for len(); added a test case for torch.fx.len()"


Differential Revision: [D25608804](https://our.internmc.facebook.com/intern/diff/D25608804)

[ghstack-poisoned]

jamesr66a commented on torch/fx/proxy.py:
```diff
@@ -153,6 +155,9 @@ def __bool__(self) -> bool:
         return self.tracer.keys(self)
 
+    def __len__(self):
+        raise RuntimeError("'builtins.len' is not supported. Replace `builtins.len` with 'torch.fx.len' and try again.")
```
Tiny nit: typically folks are going to have `len()` in their code rather than `builtins.len()`, so I think the error message here should just mention `len` without the `builtins` prefix

jamesr66a reviewed (approved):
Excellent! This is good to land after addressing the line-too-long error as well as I had one small comment

huiguoo committed:
Update on "[Issue #46210] added torch.fx.len() to provide support for len(); added a test case for torch.fx.len()"


Differential Revision: [D25608804](https://our.internmc.facebook.com/intern/diff/D25608804)

[ghstack-poisoned]

huiguoo committed:
Update on "[Issue #46210] added torch.fx.len() to provide support for len(); added a test case for torch.fx.len()"


Differential Revision: [D25608804](https://our.internmc.facebook.com/intern/diff/D25608804)

[ghstack-poisoned]

huiguoo committed:
Update on "[Issue #46210] added torch.fx.len() to provide support for len(); added a test case for torch.fx.len()"


Differential Revision: [D25608804](https://our.internmc.facebook.com/intern/diff/D25608804)

[ghstack-poisoned]
